src/cdk-experimental/scrolling/auto-size-virtual-scroll.ts

```diff
@@ -0,0 +1,228 @@
+    const firstVisibleIndex =
+        Math.min(viewport.getDataLength() - 1, Math.floor(scrollOffset / itemSize));
+    const bufferSize = Math.ceil(this._addBufferPx / itemSize);
+    const range = this._expandRange(
+        this._getVisibleRangeForIndex(firstVisibleIndex), bufferSize, bufferSize);
+
+    viewport.setRenderedRange(range);
+    viewport.setRenderedContentOffset(itemSize * range.start);
+  }
+
+  // TODO: maybe move to base class, can probably share with fixed size strategy.
+  /**
+   * Gets the visible range of data for the given start index. If the start index is too close to
+   * the end of the list it may be backed up to ensure the estimated size of the range is enough to
+   * fill the viewport.
+   * Note: must not be called if `this._viewport` is null
+   * @param startIndex The index to start the range at
+   * @return a range estimated to be large enough to fill the viewport when rendered.
+   */
+  private _getVisibleRangeForIndex(startIndex: number): Range {
+    const viewport = this._viewport!;
+    let range = {
+      start: startIndex,
+      end: startIndex +
+          Math.ceil(viewport.getViewportSize() / this._averager.getAverageItemSize())
+    };
+    const extra = range.end - viewport.getDataLength();
+    if (extra > 0) {
+      range.start = Math.max(0, range.start - extra);
+    }
+    return range;
+  }
+
+  // TODO: maybe move to base class, can probably share with fixed size strategy.
+  /**
+   * Expand the given range by the given amount in either direction.
+   * Note: must not be called if `this._viewport` is null
+   * @param range The range to expand
+   * @param expandStart The number of items to expand the start of the range by.
+   * @param expandEnd The number of items to expand the end of the range by.
+   * @return The expanded range.
+   */
+  private _expandRange(range: Range, expandStart: number, expandEnd: number): Range {
+    const viewport = this._viewport!;
+    const start = Math.max(0, range.start - expandStart);
+    const end = Math.min(viewport.getDataLength(), range.end + expandEnd);
+    return {start, end};
+  }
+
+  /** Update the viewport's total content size. */
+  private _updateTotalContentSize() {
+    const viewport = this._viewport!;
+    viewport.setTotalContentSize(viewport.getDataLength() * this._averager.getAverageItemSize());
+  }
+}
+
+/**
+ * Provider factory for `AutoSizeVirtualScrollStrategy` that simply extracts the already created
+ * `AutoSizeVirtualScrollStrategy` from the given directive.
+ * @param autoSizeDir The instance of `CdkAutoSizeVirtualScroll` to extract the
+ *     `AutoSizeVirtualScrollStrategy` from.
+ */
+export function _autoSizeVirtualScrollStrategyFactory(autoSizeDir: CdkAutoSizeVirtualScroll) {
+  return autoSizeDir._scrollStrategy;
+}
+
+
+/** A virtual scroll strategy that supports unknown or dynamic size items. */
+@Directive({
+  selector: 'cdk-virtual-scroll-viewport[autosize]',
```

Comment:
Wouldn't this selector technically catch `autosize="false"`?

Reply:
It would, but I think we just need to document that `autosize` isn't meant to take any input value. It's just a directive that provides a `VirtualScrollStrategy`